syntax = "proto3";

package flowpipe.v1;

option go_package = "github.com/hurdad/flow-pipe/gen/go/flowpipe/v1;flowpipev1";

// ============================================================
// Flow-level observability configuration (intent only)
// ============================================================
//
// - Expresses what signals a flow wants to emit
// - Does NOT define exporters, backends, or credentials
// - Runtime applies defaults, validation, and safety limits
// - Global runtime configuration always wins
//

// ------------------------------------------------------------
// OTLP transport selection
// ------------------------------------------------------------
enum OtlpTransport {
  OTLP_TRANSPORT_UNSPECIFIED = 0; // Runtime default (typically gRPC)
  OTLP_TRANSPORT_GRPC = 1;        // OTLP over gRPC
  OTLP_TRANSPORT_HTTP = 2;        // OTLP over HTTP/protobuf
}

message ObservabilityConfig {

  // ----------------------------------------------------------
  // Master enable switches
  // ----------------------------------------------------------

  bool metrics_enabled = 1;
  bool tracing_enabled = 2;
  bool logs_enabled = 3;

  // ----------------------------------------------------------
  // OTLP endpoint & transport selection
  // ----------------------------------------------------------

  string otlp_endpoint = 4;
  OtlpTransport transport = 5;

  // ==========================================================
  // Tracing configuration
  // ==========================================================
  message TracingConfig {

    // --------------------------------------------------------
    // Tracing sampling intent
    // --------------------------------------------------------
    enum TraceHint {
      TRACE_HINT_UNSPECIFIED = 0;
      TRACE_HINT_DEFAULT = 1;
      TRACE_HINT_ALWAYS = 2;
      TRACE_HINT_NEVER = 3;
    }

    TraceHint trace_hint = 1;

    // --------------------------------------------------------
    // Trace processing strategy
    // --------------------------------------------------------
    enum TraceProcessorType {
      TRACE_PROCESSOR_UNSPECIFIED = 0;
      TRACE_PROCESSOR_SIMPLE = 1;
      TRACE_PROCESSOR_BATCH = 2;
    }

    TraceProcessorType processor = 2;

    // --------------------------------------------------------
    // Batch trace processor tuning
    // --------------------------------------------------------
    message BatchConfig {
      uint32 max_queue_size = 1;
      uint32 max_export_batch_size = 2;
      uint32 schedule_delay_ms = 3;
      uint32 export_timeout_ms = 4;
    }

    BatchConfig batch = 3;

    // --------------------------------------------------------
    // Span granularity controls
    // --------------------------------------------------------
    bool stage_spans_enabled = 4;
    bool queue_spans_enabled = 5;
    bool record_spans_enabled = 6;

    // --------------------------------------------------------
    // Trace attribute verbosity
    // --------------------------------------------------------
    enum AttributeLevel {
      ATTRIBUTES_UNSPECIFIED = 0;
      ATTRIBUTES_MINIMAL = 1;
      ATTRIBUTES_STANDARD = 2;
      ATTRIBUTES_VERBOSE = 3;
    }

    AttributeLevel attribute_level = 7;

    uint32 latency_budget_ms = 8;
  }

  TracingConfig tracing = 6;

  // ==========================================================
  // Metrics configuration
  // ==========================================================
  message MetricsConfig {

    // Granularity controls (disable defaults to false)
    bool stage_metrics_disabled = 1;
    bool queue_metrics_disabled = 2;
    bool flow_metrics_disabled = 3;

    // Cost / cardinality controls
    bool latency_histograms_disabled = 4;
    bool counters_only = 5;

    // Disable jemalloc runtime metrics (if supported by runtime)
    bool jemalloc_metrics_disabled = 6;

    // Collection interval hints
    uint32 min_collection_interval_ms = 7;
    uint32 collection_interval_ms = 8;
    uint32 latency_budget_ms = 9;
  }

  MetricsConfig metrics = 7;

  // ==========================================================
  // Logging configuration
  // ==========================================================
  message LoggingConfig {

    enum LogProcessorType {
      LOG_PROCESSOR_UNSPECIFIED = 0;
      LOG_PROCESSOR_SIMPLE = 1;
      LOG_PROCESSOR_BATCH = 2;
    }

    LogProcessorType processor = 1;

    message BatchConfig {
      uint32 max_queue_size = 1;
      uint32 max_export_batch_size = 2;
      uint32 schedule_delay_ms = 3;
      uint32 export_timeout_ms = 4;
    }

    BatchConfig batch = 2;
  }

  LoggingConfig logging = 8;

  // ----------------------------------------------------------
  // Debug intent
  // ----------------------------------------------------------

  bool debug = 9;
}
